src/logic.service.ts

```diff
@@ -6,38 +6,160 @@ import type { DataCache, Pacs002, Pacs008, Pain001, Pain013 } from '@tazama-lf/f
+  const creditorMmbId = transaction.FIToFICstmrCdtTrf.CdtTrfTxInf.CdtrAgt.FinInstnId.ClrSysMmbId.MmbId;
+
+  const debtorId = generateDebtorEntityKey(tenantId, debtorOthr.Id, debtorOthr.SchmeNm.Prtry);
+  const creditorId = generateCreditorEntityKey(tenantId, creditorOthr.Id, creditorOthr.SchmeNm.Prtry);
+  const debtorAcctId = generateDebtorAccountKey(tenantId, debtorAcctOthr.Id, debtorAcctOthr.SchmeNm.Prtry, debtorMmbId);
+  const creditorAcctId = generateCreditorAccountKey(tenantId, creditorAcctOthr.Id, creditorAcctOthr.SchmeNm.Prtry, creditorMmbId);
+
+  return {
+    cdtrId: creditorId,
+    dbtrId: debtorId,
+    cdtrAcctId: creditorAcctId,
+    dbtrAcctId: debtorAcctId,
+  };
+};
+
+/**
+ * Rebuilds the DataCache object using the given endToEndId to fetch a stored Pacs008 message
+ * Always uses tenant-aware key generation with DEFAULT tenant fallback
+ *
+ * @param {string} endToEndId
+ * @param {boolean} writeToRedis
+ * @param {string} tenantId - Tenant ID (defaults to 'DEFAULT' if not provided)
+ * @param {string} id - Optional ID for logging
+ * @return {*}  {(Promise<DataCache | undefined>)}
+ */
+export const rebuildCache = async (
+  endToEndId: string,
+  writeToRedis: boolean,
+  tenantId = 'DEFAULT',
+  id?: string,
+): Promise<DataCache | undefined> => {
+  const span = apm.startSpan('db.cache.rebuild.tenant');
+  const context = 'rebuildCache()';
+  const currentPacs008 = (await cacheDatabaseManager.getTransactionPacs008(endToEndId, tenantId)) as [Pacs008[]];
+
+  const pacs008 = unwrap(currentPacs008);
+
+  if (!pacs008) {
+    loggerService.error('Could not find pacs008 transaction to rebuild dataCache with', context, id);
+    span?.end();
+    return undefined;
+  }
+
+  const cdtTrfTxInf = pacs008.FIToFICstmrCdtTrf.CdtTrfTxInf;
+
+  const { cdtrId, cdtrAcctId, dbtrId, dbtrAcctId } = parseDataCache(pacs008, tenantId);
+
+  const dataCache: DataCache = {
+    cdtrId,
+    dbtrId,
+    cdtrAcctId,
+    dbtrAcctId,
+    creDtTm: pacs008.FIToFICstmrCdtTrf.GrpHdr.CreDtTm,
+    instdAmt: {
+      amt: parseFloat(cdtTrfTxInf.InstdAmt.Amt.Amt),
+      ccy: cdtTrfTxInf.InstdAmt.Amt.Ccy,
+    },
+    intrBkSttlmAmt: {
+      amt: parseFloat(cdtTrfTxInf.IntrBkSttlmAmt.Amt.Amt),
+      ccy: cdtTrfTxInf.IntrBkSttlmAmt.Amt.Ccy,
+    },
+    xchgRate: cdtTrfTxInf.XchgRate,
+  };
+
+  if (writeToRedis) {
+    const buffer = createMessageBuffer({ DataCache: { ...dataCache } });
+
+    if (buffer) {
+      const redisTTL = configuration.redisConfig.distributedCacheTTL;
+      const tenantCacheKey = generateTenantCacheKey(tenantId, endToEndId);
+      await cacheDatabaseManager.set(tenantCacheKey, buffer, redisTTL ?? 0);
+    } else {
+      loggerService.error('[pacs008] could not rebuild redis cache');
+    }
+  }
+
+  span?.end();
+  return dataCache;
+};
+
+export const handlePain001 = async (transaction: Pain001 & { TenantId: string }, transactionType: string): Promise<void> => {
```

Comment:
The Pain001, Pain013, Pacs008, and Pacs002 interfaces already include the `TenantId` field. No need of have the & operator to 'extend' the type. This can be updated for the other three handlers as well.
```suggestion
export const handlePain001 = async (transaction: Pain001, transactionType: string): Promise<void> => {
```